docs/input_output.md

```diff
@@ -2,85 +2,279 @@
+
+> `config`: Task-specific configuration object with URL templates resolved.
+
+> `Task output`: By default, the task's return value is the next payload, or a portion of it is interpreted as the next payload as specified in the task configuration.
+
+**Cumulus Message Adapter has the following capabilities:**
+
+### 1. Retrieve a Cumulus message from S3 Bucket, or store a cumulus message to a S3 Bucket.
+
+Because of the potential size of a Cumulus message, mainly the `"payload"` field, if the message size is greater than 10000 bytes, the full message will be stored to S3 Bucket.  The message may contain a reference to an S3 Bucket and Key, as follows:
 
     {
-      "payload" : {
-        "Bucket" : "gitc-foo",
-        "Key" : "bar/baz"
+      "cumulus_meta": {...},
+      "replace": {
+        "Bucket": "gitc-foo",
+        "Key": "bar/baz"
+        }
+    }
+
+When a workflow task receives such message, the Cumulus Message Adapter is responsible for fetching the JSON message document and pass it to the task.  When a task output message is too big, the Cumulus Message Adapter will store the message to S3 Bucket under $.cumulus_meta.buckets.internal, and return a new message with S3 reference as above example.
+
+### 2. Resolve URL templates in the task configuration
+
+In the workflow configuration, each task has its own configuration, and it can use URL template as a value for achieve simplicity or for values only available at execution time.  When each task executes, it is expected to resolve URL templates found in its configuration against the entire Cumulus message. The Cumulus Message Adapter solves the URL templates and then passes message to next task. For example, a task has configuration in the workflow:
+
+    Discovery:
+        config:
+          useQueue: true
+          stack: '$.cumulus_meta.stack'
+          provider: '$.meta.provider'
+          inlinestr: 'prefix{meta.foo}suffix',
+          array: '{[$.meta.foo]}',
+          object: '{{$.meta}}'
+
+The task configuration in the message (only partial message is shown here):
+
+    {
+      "cumulus_meta": {
+        "stack": "foo-cumulus",
+        ....
+      },
+      "meta": {
+        "foo": "bar",
+        "provider": {
+          "id": "FOO_DAAC",
+          "anykey": "anyvalue"
+        },
+        ...
+      },
+      "workflow_config": {
+        "Discovery": {
+          "useQueue": true,
+          "stack": "{{$.cumulus_meta.stack}}",
+          "object": "{{$.meta.provider}}",
+          "inlinestr": "prefix{meta.foo}suffix",
+          "array": "{[$.meta.foo]}"
+        },
+        ...
       }
     }
 
-When a Task receives such a payload, it is responsible for fetching the JSON document at the given key and replacing `"payload"` with its contents.  See the (EventSource)[../lib/event-source.js] implementation for examples.
+Into this as part of the message passed to task:
```

Comment:
This part doesn't seem clear to me - where is the object below coming from and where is it going?